src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,238 @@
+            } else if (index == size) {
+                add(value);
+            } else {
+                addToMiddle(newNode, index);
+                size++;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T el: list) {
+            add(el);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> newNode = null;
+        if (index >= 0 && index < size) {
+            if (firstHalf(index)) {
+                int indexFirstHalf = index + ONE;
+                newNode = head;
+                newNode = firstHalfFindNode(newNode, indexFirstHalf);
+            } else {
+                newNode = tail;
+                int indexFromEnd = size - index - ONE;
+                newNode = secondHalfFindNode(newNode, indexFromEnd);
+            }
+            return newNode.item;
+        } else {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        if (node != null) {
+            T item = node.item;
+            node.item = value;
+            return item;
+        } else {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = getNode(index);
+        if (index == NOT_FOUND) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        if (head == tail) {
+            final T removedItem = head.item;
+            head = null;
+            tail = null;
+            size = 0;
+            return removedItem;
+        }
+        if (index == 0) {
+            Node<T> nextNode = node.next;
+            nextNode.prev = null;
+            head = nextNode;
+            size--;
+        } else if (index == size - ONE) {
+            Node<T> prevNode = node.prev;
+            prevNode.next = null;
+            tail = prevNode;
+            size--;
+        } else {
+            Node<T> prevNode = node.prev;
+            Node<T> nextNode = node.next;
+            prevNode.next = nextNode;
+            nextNode.prev = prevNode;
+            size--;
+        }
+
+        return node.item;
     }
 
     @Override
     public boolean remove(T object) {
+        int index = getNode(object);
+        if (index != NOT_FOUND) {
+            remove(index);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isValidIndex(int index) {
+        if (index >= 0 && index <= size) {
+            return true;
+        } else {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+    }
```

Comment:
we can make it more simple
```suggestion
    private void checkIndex(int index) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }
    }
```